common/src/main/java/com/microsoft/identity/common/internal/logging/Logger.java

```diff
@@ -116,71 +119,395 @@ public void setExternalLogger(final ILoggerCallback externalLogger) {
+     * @param message       The message to log.
      */
-    public static void verbosePII(final String tag, final String correlationID, final String message) {
-        getInstance().log(tag, LogLevel.VERBOSE, correlationID, message, null, true);
+    public static void verbosePII(final String tag,
+                                  @Nullable final String correlationID,
+                                  @Nullable final String message) {
+        getInstance().log(
+                tag,
+                LogLevel.VERBOSE,
+                correlationID,
+                message,
+                null,
+                true
+        );
     }
 
     /**
-     * TODO. Need to discuss on how to keep the correlationID. CorrelationID should be per request => need to sync with Telemetry implementation
+     * Log a method entry, with optional parameters. This method only writes output if the
+     * {@link LogLevel} is {@link LogLevel#VERBOSE} and PII logging is enabled.
      *
+     * @param tag    Used to identify the source of a log message. It usually identifies the class
+     *               or activity where the log call occurs.
+     * @param method The method which is being entered.
+     * @param args   The arguments to the method being entered.
+     */
+    public static void entering(final String tag,
+                                final String method,
+                                @Nullable final Object... args) {
+        verbosePII(tag, formatMethodArgs(method, true, args));
+    }
+
+    /**
+     * Log a method exit, with an optional exit value. This method only writes output if the
+     * {@link LogLevel} is {@link LogLevel#VERBOSE} and PII logging is enabled.
+     *
+     * @param tag    Used to identify the source of a log message. It usually identifies the class
+     *               or activity where the log call occurs.
+     * @param method The method which is being exited.
+     * @param arg    The method's return-value.
+     */
+    public static void exiting(final String tag,
+                               final String method,
+                               @Nullable final Object... arg) {
+        verbosePII(tag, formatMethodArgs(method, false, arg));
+    }
+
+    private static String formatMethodArgs(final String method, final boolean entering, final Object[] args) {
+        final String argSeparator = ", ";
+
+        // Take the form of '^methodName(' or '$methodName(' [enter vs. exit]
+        String output = (entering ? "^" : "$") + method + "(";
+
+        if (null != args && 0 != args.length) {
+            for (final Object arg : args) {
+                if (null != arg) {
+                    if (entering) {
+                        final Class<?> clazz = arg.getClass();
+                        output += clazz.getSimpleName() + "=" + arg.toString() + argSeparator;
+                    } else {
+                        output += "result=" + arg.toString();
```

Comment:
This looks to me like a requirement to implement toString properly on all our domain objects.  So that our verbose logging tells us what we need to know.